server/src/main/java/org/elasticsearch/search/query/PartialHitCountCollector.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.query;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.FilterLeafCollector;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.TotalHitCountCollector;
+
+import java.io.IOException;
+
+/**
+ * Extension of {@link TotalHitCountCollector} that supports early termination of total hits counting based on a provided threshold.
+ * Note that the total hit count may be retrieved from {@link org.apache.lucene.search.Weight#count(LeafReaderContext)},
+ * in which case early termination is only applied to the leaves that do collect documents.
+ */
+class PartialHitCountCollector extends TotalHitCountCollector {
+
+    private final int totalHitsThreshold;
+    // we could reuse the counter that TotalHitCountCollector has and exposes through getTotalHits(),
+    // but that would make us early terminate also when retrieving count from Weight#count and would
+    // cause a behaviour that's difficult to explain and test.
+    private int numCollected = 0;
+    private boolean earlyTerminated;
+
+    PartialHitCountCollector(int totalHitsThreshold) {
+        this.totalHitsThreshold = totalHitsThreshold;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        // Does not need scores like TotalHitCountCollector (COMPLETE_NO_SCORES), but not exhaustive as it early terminates.
+        return ScoreMode.TOP_DOCS;
+    }
+
+    @Override
+    public LeafCollector getLeafCollector(LeafReaderContext context) throws IOException {
+        if (numCollected >= totalHitsThreshold) {
+            earlyTerminateIfNeeded();
+        }
+        return new FilterLeafCollector(super.getLeafCollector(context)) {
```

Comment:
Just for my clarity, the idea is we do not consider it early termination if we are able to get the count from `Weight#count` in this case and an internal `CollectionTerminatedException` is thrown?

Reply:
yes, retrieving from Weight#count goes segment by segment, hence some segments may have to collect, while some others not. We only do early termination when we collect. This is not changing any current behaviour. When count is retrieved from the weight a `CollectionTerminatedException` is thrown for that segment (from TotalHitCountCollector) to signal that we can move on to the next segment.